deps/internal/cmds/get/topology.go

```diff
@@ -147,13 +148,134 @@ func topology(ctx context.Context, searchService tracker.SearchServiceClient, re
 	return result, nil
 }
 
+func getAllPaths(ctx context.Context, searchService tracker.SearchServiceClient, request *tracker.SearchRequest, destinationNodeKey string) ([][]*schema.Module, error) {
+	// Unclear as to what's the significant advantage we are achieving with using DFS over BFS
+	// As it's written right now, we do wait until the search is fully exhausted (the for loop below) anyway,
+	// before we begin to build the topology
+	call, err := searchService.DepthFirstSearch(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := call.Send(request); err != nil {
+		return nil, err
+	}
+	nodes := make(map[string]*schema.Module) // key: Module, value: Metadata of the module
+	edges := make(map[string][]string)       // key: Module, value: adjacency list of the module
+
+	// loop until all the responses are received from the server
+	for resp, err := call.Recv(); true; resp, err = call.Recv() {
+		if err == io.EOF {
+			break
+		} else if err != nil {
+			return nil, err
+		}
+
+		var source *tracker.DependencyRequest
+		var items []*tracker.Dependency
+
+		if source = resp.GetRequest().GetDependentsOf(); source != nil {
+			items = resp.GetDependents()
+		} else if source = resp.GetRequest().GetDependenciesOf(); source != nil {
+			items = resp.GetDependencies()
+		}
+
+		if source == nil {
+			return nil, fmt.Errorf("source module not included")
+		}
+
+		module := requestToModule(source)
+		moduleKey := key(module)
+		if _, ok := nodes[moduleKey]; !ok {
+			nodes[moduleKey] = module
+		}
+
+		dependencyKeys := make([]string, len(items))
+
+		for i, dependency := range items {
+			dependencyKey := key(dependency.Module)
+
+			if _, ok := nodes[dependencyKey]; !ok {
+				nodes[dependencyKey] = dependency.Module
+			}
+
+			dependencyKeys[i] = dependencyKey
+		}
+
+		edges[moduleKey] = dependencyKeys
+	}
+
+	var rootKey string
+	if req := request.GetDependentsOf(); req != nil {
+		rootKey = keyForRequest(req)
+	} else if req := request.GetDependenciesOf(); req != nil {
+		rootKey = keyForRequest(req)
+	} else {
+		return nil, fmt.Errorf("failed to determine root key for topological paths")
+	}
+
+	stack := [][]string{{rootKey}}
+	result := [][]*schema.Module{{}}
+
+	for length := len(stack); length > 0; length = len(stack) {
+		next := make([][]string, 0)
+
+		// Pop
+		path := stack[length-1]
+		stack = stack[0:(length - 1)]
```

Comment:
Since this is mostly new code, you might look into rebasing on `main`. I added a `LinkedList` implementation that can behave as a stack. It avoids a lot of the array slicing that a lot of implementations require.

Reply:
Updated the implementation with LinkedList.